data-products/src/freestar_revenue_reporting/freestar_reporting_flow.py

```diff
@@ -0,0 +1,246 @@
+                        item[new_key] = item.pop(key)
+            
+            # Append the processed data to the combined_data list
+            combined_data.extend(response_json['data'])
+
+            # Check if there are more records to retrieve
+            if len(response_json['data']) < 2000:
+                break  # Stop if there are no more records
+
+            # Increment the offset for the next page
+            request_payload["query"]["offset"] += 2000
+            page_number += 1
+
+            # Add a 5-second delay before fetching the next page
+            time.sleep(5)
+
+        except requests.exceptions.RequestException as e:
+            print(f"Error: {e}")
+
+        except json.JSONDecodeError as e:
+            print(f"Error decoding JSON response: {e}")
+
+    print(f"Total records retrieved: {len(combined_data)}")
+
+    # Convert the combined data to Parquet format and save it
+    output_parquet_filename = "data_combined.parquet"
+    save_data_to_parquet(combined_data, output_parquet_filename)
+    print("Combined data saved to", output_parquet_filename)
+
+# Define table to load
+snowflake_table = "freestar_daily_extracts"
+
+# Define SQL statements
+create_table_sql = f"""
+CREATE TABLE IF NOT EXISTS {snowflake_table} (
+    record_date DATE,
+    network STRING,
+    url STRING,
+    utm_campaign STRING,
+    utm_content STRING,
+    utm_medium STRING,
+    utm_source STRING,
+    utm_term STRING,
+    site_domain STRING,
+    ad_unit STRING,
+    size STRING,
+    country_code STRING,
+    device_type STRING,
+    device_os STRING,
+    browser STRING,
+    integration_partner STRING,
+    impressions INTEGER,
+    net_revenue INTEGER,
+    net_cpm FLOAT
+);
+"""
+
+format_file_sql = f"""
+CREATE OR REPLACE FILE FORMAT {snowflake_table}_format
+TYPE = parquet;
+"""
+
+create_stage_sql = f"""
+CREATE OR REPLACE STAGE {snowflake_table}_stage
+FILE_FORMAT = {snowflake_table}_format;
+"""
+
+put_parquet_sql = f"""
+PUT file://{output_parquet_filename} @{snowflake_table}_stage
+"""
+
+load_sql = f"""
+COPY INTO {snowflake_table}
+    FROM (SELECT $1:record_date::DATE,
+                 $1:network::STRING,
+                 $1:url::STRING,
+                 $1:utm_campaign::STRING,
+                 $1:utm_content::STRING,
+                 $1:utm_medium::STRING,
+                 $1:utm_source::STRING,
+                 $1:utm_term::STRING,
+                 $1:site_domain::STRING,
+                 $1:ad_unit::STRING,
+                 $1:size::STRING,
+                 $1:country_code::STRING,
+                 $1:device_type::STRING,
+                 $1:device_os::STRING,
+                 $1:browser::STRING,
+                 $1:integration_partner::STRING,
+                 $1:impressions::INTEGER,
+                 $1:net_revenue::INTEGER,
+                 $1:net_cpm::FLOAT
+        FROM @{snowflake_table}_stage/data_combined.parquet
+        );
+"""
```

Comment:
Also I noticed the table is created with `IF NOT EXISTS`, so I assume it might have existed already. Then some day we might inadvertently run this script twice and get duplicates in this snowflake table. Do you know is there a way to make this COPY idempotent, so it wouldn't matter if we run it multiple times, it would not create duplicates?

Reply:
Great point on the idempotent COPY